src/D2L.CodeStyle.Analyzers/Immutability/ImmutabilityExceptionInheritanceAnalyzer.cs

```diff
@@ -22,70 +21,135 @@ public sealed class ImmutabilityExceptionInheritanceAnalyzer : DiagnosticAnalyze
+			// involves setting the exceptions to this set, i.e. we will never
+			// suggest to *add* new kinds of exceptions.
+			var maximalExceptions = new HashSet<string>( directExceptions );
 
-					var location = GetLocationOfClassIdentifierAndGenericParameters( root );
-					var diagnostic = Diagnostic.Create(
-						Diagnostics.ImmutableExceptionInheritanceIsInvalid,
-						location,
-						inheritedExceptions.Key.Name,
-						suggestedFix
-					);
+			ISymbol aSuperTypeWithFewerAllowedExceptions = null;
 
-					context.ReportDiagnostic( diagnostic );
+			foreach( var inheritedExceptions in allInheritedExceptions ) {
+				if( !directExceptions.IsSubsetOf( inheritedExceptions.Value ) ) {
+					aSuperTypeWithFewerAllowedExceptions = inheritedExceptions.Key;
+					maximalExceptions.IntersectWith( inheritedExceptions.Value );
 				}
 			}
+
+			if ( aSuperTypeWithFewerAllowedExceptions == null ) {
+				// We didn't find anything wrong.
+				return;
+			}
+
+			var location = GetImmutableAttributeSyntax(
+				context.SemanticModel,
+				immutableAttribute,
+				declSyntax
+			).GetLocation();
+
+			var fixInfo = GetInfoForFix( maximalExceptions );
+
+			var diagnostic = Diagnostic.Create(
+				Diagnostics.ImmutableExceptionInheritanceIsInvalid,
+				location,
+				messageArgs: new[] { aSuperTypeWithFewerAllowedExceptions.Name },
+				properties: fixInfo
+			);
+
+			context.ReportDiagnostic( diagnostic );
 		}
 
-		private Location GetLocationOfClassIdentifierAndGenericParameters(
+		/// <summary>
+		/// Format necessary context for the code fix
+		/// </summary>
+		/// <param name="maximalExceptions">The largest set of exception kinds
+		/// we are allowed to have</param>
+		/// <returns>Object that can be passed to the code fix</returns>
+		private ImmutableDictionary<string, string> GetInfoForFix(
+			IEnumerable<string> maximalExceptions
+		) {
+			var sortedMaximalExceptions = maximalExceptions
+				.OrderBy( v => v );
+
+			var serializedExceptions = string.Join(
+				",",
+				sortedMaximalExceptions
+			);
+
+			return ImmutableDictionary.Create<string, string>()
+				.Add( "excepts", serializedExceptions )
+				.ToImmutableDictionary();
+		}
+
+		private static AttributeSyntax GetImmutableAttributeSyntax(
+			SemanticModel model,
+			INamedTypeSymbol immutableAttribute,
 			TypeDeclarationSyntax decl
 		) {
-			var location = decl.Identifier.GetLocation();
-
-			if( decl.TypeParameterList != null ) {
-				location = Location.Create(
-					decl.SyntaxTree,
-					TextSpan.FromBounds(
-						location.SourceSpan.Start,
-						decl.TypeParameterList.GetLocation().SourceSpan.End
-					)
-				);
+			var attrs = decl.AttributeLists
+				.SelectMany( al => al.Attributes );
+
+			foreach( var attr in attrs ) {
+				var attrType = model.GetSymbolInfo( attr )
+					.Symbol // the symbol for one of the attribute constructors
+					.ContainingType; // the symbol for the attributes type
+
+				if( attrType == immutableAttribute ) {

```

Comment:
Not necessarily an issue with this PR, but in other places in the analyzers we [support having multiple `Immutable` attributes](https://github.com/Brightspace/D2L.CodeStyle/blob/master/src/D2L.CodeStyle.Analyzers/Immutability/ImmutableHelpers.cs#L66). However, `[Immutable]` isn't defined with `AllowMultiple` anyways, so I don't think it's an issue currently.

Reply:
Yeah, those helpers are pretty insane in general. `[Immutable]` is definitely not `AllowMultiple` by choice.
This PR is actually to help me land #411 which is part of a larger effort to simplify this stuff.

(I also really hate those helpers because they are built on top of string manipulation... too much framework to enable doing the wrong thing...)

Reply:
So ya good catch. I hope to give those parts of the code some good hard love soon...